buffer/ring_growing.go

```diff
@@ -16,57 +16,72 @@ limitations under the License.
 
 package buffer
 
+import "sync/atomic"
+
 // RingGrowing is a growing ring buffer.
 // Not thread safe.
 type RingGrowing struct {
 	data     []interface{}
-	n        int // Size of Data
-	beg      int // First available element
-	readable int // Number of data items available
+	n        int32 // Size of Data
+	beg      int32 // First available element
+	readable int32 // Number of data items available
 }
 
 // NewRingGrowing constructs a new RingGrowing instance with provided parameters.
 func NewRingGrowing(initialSize int) *RingGrowing {
 	return &RingGrowing{
 		data: make([]interface{}, initialSize),
-		n:    initialSize,
+		n:    int32(initialSize),
 	}
 }
 
 // ReadOne reads (consumes) first item from the buffer if it is available, otherwise returns false.
 func (r *RingGrowing) ReadOne() (data interface{}, ok bool) {
-	if r.readable == 0 {
+	oldReadable := atomic.LoadInt32(&r.readable)
+	oldN := atomic.LoadInt32(&r.n)
+	if oldReadable == 0 {
+		return nil, false
+	}
+	if !atomic.CompareAndSwapInt32(&r.readable, oldReadable, oldReadable-1) {
```

Comment:
Suppose `r.readable` is 100, it's possible that two goroutines are trying to execute this `compareAndSwapInt32` at the same time and only one gorouting will succeed. This means even we will return `nil` even if `r.readable > 0`.

I haven't write go code for a long time, but can we use `atomic.Xadd` here which should be much easier.